crates/scmi/src/devices/iio.rs

```diff
@@ -0,0 +1,851 @@
+                "No {} channel found in {}",
+                &self.channel, &self.path
+            ));
+        }
+        axes.sort_by(|a1, a2| a1.path.cmp(&a2.path));
+        self.axes = axes;
+        Ok(())
+    }
+
+    fn unit(&self) -> u8 {
+        UNIT_MAPPING
+            .iter()
+            .find(|mapping| mapping.channel == self.channel)
+            .map_or(scmi::SENSOR_UNIT_UNSPECIFIED, |mapping| mapping.unit)
+    }
+
+    fn unit_exponent(&self, axis_index: u32) -> i8 {
+        let axis: &Axis = self.axes.get(axis_index as usize).unwrap();
+        axis.unit_exponent + axis.custom_exponent
+    }
+
+    fn number_of_axes(&self) -> u32 {
+        if self.scalar {
+            0
+        } else {
+            self.axes.len() as u32
+        }
+    }
+
+    fn axis_name_prefix(&self) -> String {
+        let channel = &self.channel;
+        let in_prefix = "in_";
+        let out_prefix = "out_";
+        let name: &str = if channel.starts_with(in_prefix) {
+            channel.strip_prefix(in_prefix).unwrap()
+        } else if channel.starts_with(out_prefix) {
+            channel.strip_prefix(out_prefix).unwrap()
+        } else {
+            &channel[..]
+        };
+        let len = min(name.len(), MAX_SIMPLE_STRING_LENGTH - 1);
+        String::from(&name[..len])
+    }
+
+    fn reading_get(&mut self) -> DeviceResult {
+        let mut result = vec![];
+        for axis in &self.axes {
+            let value = self.read_axis(axis)?;
+            result.push(MessageValue::Unsigned((value & 0xFFFFFFFF) as u32));
+            result.push(MessageValue::Unsigned((value >> 32) as u32));
+            result.push(MessageValue::Unsigned(0));
+            result.push(MessageValue::Unsigned(0));
+        }
+        Result::Ok(result)
+    }
+}
+
+fn read_number_from_file<F: FromStr>(path: &Path) -> Result<Option<F>, ScmiDeviceError> {
+    let exists = path.try_exists();
+    if exists.is_err() || exists.as_ref().unwrap() == &false {
+        let raw = path.ends_with("_raw");
+        // TODO: Is this formatted unnecessarily if debug logging is not enabled?
+        let message = format!(
+            "IIO {} file {} not found{}",
+            if raw { "value" } else { "data" },
+            path.to_string_lossy(),
+            match exists.as_ref() {
+                Ok(_) => String::from(""),
+                Err(error) => String::from(": ") + &error.to_string(),
+            }
+        );
+        if raw {
+            error!("{message}");
+            return Err(ScmiDeviceError::GenericError);
+        } else {
+            debug!("{message}");
+            return Ok(None);
+        }
+    }
+    let file_result = fs::read_to_string(path);
+    if let Err(error) = file_result {
+        error!(
+            "Failed to read IIO data from {}: {}",
+            path.to_string_lossy(),
+            error
+        );
+        return Err(ScmiDeviceError::GenericError);
+    }
+    let string = &file_result.unwrap();
```

Comment:
this sounds like a good if let else use case :)

Reply:
Well, as you are returning Result<(), T>, one could just propagate the error and not manually handle every single case. The error types should already give you enough information about the failure anyway

Reply:
The other option is to use something like
```rust
    match fs::read_to_string("path").map(|s| s.parse::<u32>()) {
        Ok(value) => Ok(Some(value)),
        Err(err) => {
            error!("Failed to read/parse the value from path");
            ....
        }
    }
```

Reply:
I've rewritten the function based on your other comment. Is it better now?
One thing to keep on mind is that it's important to be clear with logging and error messages because every bit of information may be important in case of non-reproducible errors, missing access to the given system, in logs obtained from users.